File: python/sample_05_balance.py
"""
Sample 5 — Lấy số dư tài khoản (Account Balance)
===================================================
Kiểm tra khả năng giao dịch trước khi đặt lệnh.

Luồng:
  1. Client gọi endpoint balances theo accountId
  2. API trả về available, onHold, limits, settlement
  3. Tính khả năng mua/bán thực tế theo nghiệp vụ
  4. Nếu không đủ điều kiện thì chặn thao tác đặt lệnh
  5. Nếu đủ điều kiện thì cho phép đi tiếp sang order flow
"""

from ssi_sdk import Auth, Trading, Config
from auth_helper import ensure_auth

config = Config(
    client_id="<CLIENT_ID>",
    api_key="<API_KEY>",
    api_secret="<API_SECRET>",
    private_key=("<PRIVATE_KEY_CONTENT>"),
    log_level="DEBUG",
)

ACCOUNT_NO = "<ACCOUNT_NO>"

with Auth(config) as auth:
    ensure_auth(auth, otp="203081")

    with Trading(auth) as trading:
        # --- Bước 1: Lấy danh sách tài khoản ---
        accounts = trading.account.get_account_info()
        print("Danh sách tài khoản:")
        for acc in accounts:
            print(f"  - {acc.account_no} ({acc.account_type.value})")

        # --- Bước 2: Lấy số dư tài khoản Equity ---
        print(f"\n--- Số dư tài khoản Equity: {ACCOUNT_NO} ---")
        balance = trading.portfolio.get_equity_balance(ACCOUNT_NO)
        print(f"  Tiền mặt khả dụng  : {balance.available_cash:>15,.0f}")
        print(f"  Tổng nợ            : {balance.total_debt:>15,.0f}")
        print(f"  Mua T0/T1/T2       : {balance.buy_t0:>12,.0f} / {balance.buy_t1:>12,.0f} / {balance.buy_t2:>12,.0f}")
        print(f"  Bán T0/T1/T2       : {balance.sell_t0:>12,.0f} / {balance.sell_t1:>12,.0f} / {balance.sell_t2:>12,.0f}")

        # --- Bước 3: Kiểm tra sức mua tối đa cho một mã ---
        print("\n--- Sức mua/bán tối đa: SSI ---")
        max_bs = trading.trading.get_max_buy_sell(ACCOUNT_NO, "SSI", 26000)
        print(f"  Max mua : {max_bs.max_buy_quantity:>10} cổ phiếu")
        print(f"  Max bán : {max_bs.max_sell_quantity:>10} cổ phiếu")
        print(f"  Sức mua : {max_bs.purchase_power:>15,.0f}")

        # --- Bước 4: Logic kiểm tra trước khi đặt lệnh ---
        desired_quantity = 100
        desired_price = 26000
        required_amount = desired_quantity * desired_price

        if balance.available_cash >= required_amount:
            print(f"\n✓ Đủ điều kiện: cần {required_amount:,.0f}, có {balance.available_cash:,.0f}")
            print("  → Cho phép đặt lệnh mua.")
        else:
            print(f"\n✗ Không đủ: cần {required_amount:,.0f}, chỉ có {balance.available_cash:,.0f}")
            print("  → Chặn đặt lệnh.")

        # --- Bước 5: Xem vị thế hiện có ---
        print(f"\n--- Vị thế cổ phiếu ({ACCOUNT_NO}) ---")
        positions = trading.portfolio.get_equity_positions(ACCOUNT_NO)
        for pos in positions:
            print(
                f"  {pos.symbol:<10} | SL: {pos.quantity:>8} | "
                f"Bán được: {pos.sellable_quantity:>8} | Giá vốn: {pos.cost_price:>10,.0f}"
            )

File: go/sample_05_balance.go
/*
Sample 5 — Lấy số dư tài khoản (Account Balance)
===================================================
Kiểm tra khả năng giao dịch trước khi đặt lệnh.

Luồng:
 1. Client gọi endpoint balances theo accountId
 2. API trả về available, onHold, limits, settlement
 3. Tính khả năng mua/bán thực tế theo nghiệp vụ
 4. Nếu không đủ điều kiện thì chặn thao tác đặt lệnh
 5. Nếu đủ điều kiện thì cho phép đi tiếp sang order flow
*/
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os"

	"github.com/SSI-Securities-Inc/ssi-sdk-go/v3/auth"
	"github.com/SSI-Securities-Inc/ssi-sdk-go/v3/ssi"
)

type appConfig struct {
	ClientID          string `json:"client_id"`
	APIKey            string `json:"api_key"`
	APISecret         string `json:"api_secret"`
	PrivateKey        string `json:"private_key"`
	OTP               string `json:"otp"`
	EquityAccount     string `json:"equity_account"`
	DerivativeAccount string `json:"derivative_account"`
	LogLevel          string `json:"log_level"`
}

func loadConfig() (*ssi.Config, string, string) {
	config := ssi.NewConfig("<CLIENT_ID>")
	config.APIKey = "<API_KEY>"
	config.APISecret = "<API_SECRET>"
	config.PrivateKey = "<PRIVATE_KEY_CONTENT>"
	accountNo := "<ACCOUNT_NO>"
	otp := "<OTP>"

	for _, path := range []string{"../config.json", "config.json"} {
		data, err := os.ReadFile(path)
		if err != nil {
			continue
		}
		var appCfg appConfig
		if err := json.Unmarshal(data, &appCfg); err == nil {
			if appCfg.ClientID != "" {
				config.ClientID = appCfg.ClientID
			}
			if appCfg.APIKey != "" {
				config.APIKey = appCfg.APIKey
			}
			if appCfg.APISecret != "" {
				config.APISecret = appCfg.APISecret
			}
			if appCfg.PrivateKey != "" {
				config.PrivateKey = appCfg.PrivateKey
			}
			if appCfg.EquityAccount != "" {
				accountNo = appCfg.EquityAccount
			}
			if appCfg.OTP != "" {
				otp = appCfg.OTP
			}
			break
		}
	}
	return config, accountNo, otp
}

func main() {
	config, accountNo, otp := loadConfig()

	auth := ssi.NewAuth(config)
	defer auth.Close()

	ensureAuth(auth, otp)

	t := ssi.NewTrading(auth)

	// --- Bước 1: Lấy danh sách tài khoản ---
	accounts, err := t.Account.GetAccountInfo()
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println("Danh sách tài khoản:")
	for _, acc := range accounts {
		fmt.Printf("  - %s (%s)\n", acc.AccountNo, acc.AccountType)
	}

	// --- Bước 2: Lấy số dư tài khoản Equity ---
	fmt.Printf("\n--- Số dư tài khoản Equity: %s ---\n", accountNo)
	balance, err := t.Portfolio.GetEquityBalance(accountNo)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("  Tiền mặt khả dụng : %.0f\n", balance.AccountBalance)

	// --- Bước 3: Kiểm tra sức mua tối đa cho một mã ---
	fmt.Println("\n--- Sức mua/bán tối đa: SSI ---")
	price := 26000.0
	maxBS, err := t.Trading.GetMaxBuySell(accountNo, "SSI", &price)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("  Max mua  : %d cổ phiếu\n", maxBS.MaxBuyQuantity)
	fmt.Printf("  Max bán  : %d cổ phiếu\n", maxBS.MaxSellQuantity)

	// --- Bước 4: Logic kiểm tra trước khi đặt lệnh ---
	desiredQuantity := 100
	desiredPrice := 26000.0
	requiredAmount := float64(desiredQuantity) * desiredPrice

	if balance.AccountBalance >= requiredAmount {
		fmt.Printf("\n✓ Đủ điều kiện: cần %.0f, có %.0f\n", requiredAmount, balance.AccountBalance)
		fmt.Println("  → Cho phép đặt lệnh mua.")
	} else {
		fmt.Printf("\n✗ Không đủ: cần %.0f, chỉ có %.0f\n", requiredAmount, balance.AccountBalance)
		fmt.Println("  → Chặn đặt lệnh.")
	}

	// --- Bước 5: Xem vị thế hiện có ---
	fmt.Printf("\n--- Vị thế cổ phiếu (%s) ---\n", accountNo)
	positions, err := t.Portfolio.GetEquityPositions(accountNo)
	if err != nil {
		log.Fatal(err)
	}
	for _, pos := range positions {
		fmt.Printf("  %-10s | SL: %d | Bán được: %d | Giá vốn: %.0f\n",
			pos.Symbol, pos.Quantity, pos.SellableQuantity, pos.CostPrice)
	}

	// --- Response Summary ---
	fmt.Println("\n[Response] accounts|avail_cash|max_buy_qty|max_sell_qty|positions")
	fmt.Printf("%d|%.0f|%d|%d|%d\n", len(accounts), balance.AccountBalance, maxBS.MaxBuyQuantity, maxBS.MaxSellQuantity, len(positions))
	if len(positions) > 0 {
		p := positions[0]
		fmt.Println("[Response:first_pos] symbol|quantity|sellable|cost_price")
		fmt.Printf("%s|%d|%d|%.0f\n", p.Symbol, p.Quantity, p.SellableQuantity, p.CostPrice)
	}
}

// ── Token cache helper ──────────────────────────────────────────────────────

const sharedTokenFile = "../shared_token.json"
const tokenCacheFile = "token_cache.json"

func loadToken() *auth.Token {
	for _, file := range []string{sharedTokenFile, tokenCacheFile} {
		data, err := os.ReadFile(file)
		if err != nil {
			continue
		}
		var token auth.Token
		if err := json.Unmarshal(data, &token); err != nil {
			continue
		}
		if token.AccessToken != "" {
			return &token
		}
	}
	return nil
}

func saveToken(token *auth.Token) {
	data, err := json.MarshalIndent(token, "", "  ")
	if err != nil {
		log.Printf("Lỗi khi serialize token: %v", err)
		return
	}
	if err := os.WriteFile(tokenCacheFile, data, 0600); err != nil {
		log.Printf("Lỗi khi lưu token: %v", err)
		return
	}
	fmt.Printf("Token đã lưu vào %s\n", tokenCacheFile)
}

func ensureAuth(a *ssi.Auth, otp string) {
	token := loadToken()

	if token != nil {
		// Luôn set token từ cache vào manager trước
		a.TokenManager.SetToken(token)

		if !a.TokenManager.IsTokenExpired() {
			fmt.Println("Token còn hạn, dùng token từ file.")
			return
		}

		if !a.TokenManager.IsRefreshTokenExpired() {
			fmt.Println("Access token hết hạn, đang refresh...")
			newToken, err := a.Refresh()
			if err != nil {
				log.Printf("Refresh thất bại (%v), đang authenticate lại...", err)
			} else {
				saveToken(newToken)
				fmt.Println("Refresh token thành công.")
				return
			}
		}
	}

	fmt.Println("Không tìm thấy token hợp lệ, đang authenticate...")
	newToken, err := a.Authenticate(otp)
	if err != nil {
		log.Fatalf("Authenticate thất bại: %v", err)
	}
	saveToken(newToken)
	fmt.Println("Authenticate thành công.")
}
